// ea-lib/lib/eventtarget.js
var EventTarget = function() {
  this.listeners = {};
};

EventTarget.prototype.listeners = null;
EventTarget.prototype.addEventListener = function(type, callback) {
  if (!(type in this.listeners)) {
    this.listeners[type] = [];
  }
  this.listeners[type].push(callback);
};
EventTarget.prototype.on = EventTarget.prototype.addEventListener;

EventTarget.prototype.removeEventListener = function(type, callback) {
  if (!(type in this.listeners)) {
    return;
  }
  var stack = this.listeners[type];
  for (var i = 0, l = stack.length; i < l; i++) {
    if (stack[i] === callback){
      stack.splice(i, 1);
      return;
    }
  }
};

EventTarget.prototype.dispatchEvent = function(event, target) {
  if(typeof(event) == "string") {
    event = {
      type: event,
      defaultPrevented: false,
      target: this
    };
  }
  if(target) {
    Object.assign(event, target);
  }
  if (!(event.type in this.listeners)) {
    return true;
  }
  var stack = this.listeners[event.type];

  for (var i = 0, l = stack.length; i < l; i++) {
    var fn = stack[i];
    if(fn.call) {
      fn.call(this, event);
    } else {
      fn(event);
    }
  }
  return !event.defaultPrevented;
};
EventTarget.prototype.emit = EventTarget.prototype.dispatchEvent;


module.exports = EventTarget;
